singularity-eos/eos/eos_powermg.hpp

```diff
@@ -0,0 +1,440 @@
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real value = (1.0 - _G0 * eta / 2.0) * _K0toK40[0];
+  Real sum = 0.0;
+  for (int ind = _M; ind >= 1; ind--) {
+    sum = sum * eta + ind * _K0toK40[ind] * eta;
+  }
+  value = value * sum;
+  value = value + robust::ratio(_HugPressureFromDensity(rho), eta);
+  value = value + _G0 * _G0 * _rho0 * (sie - AllHugInternalEnergyFromDensity(rho));
+  value = robust::ratio(_rho0, rho) * value;
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::BulkModulusFromDensityInternalEnergy(
+    const Real rho, const Real sie, Real *lambda) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real value;
+  if (eta <= 0.0) {
+    if (eta < _Pmin / _K0toK40[0]) {
+      value = _G0 * _G0 * _rho0 *
+              (sie - _E0 + _Pmin / _rho0 * (_Pmin / _K0toK40[0] / 2.0 - eta));
+    } else {
+      value =
+          _K0toK40[0] + _G0 * _G0 * _rho0 * (sie - _E0 - _K0toK40[0] * eta * eta / 2.0);
+    }
+    value = robust::ratio(_rho0, rho) * value;
+  } else {
+    value = _compBulkModulusFromDensityInternalEnergy(rho, sie);
+  }
+  return value;
+}
+
+PORTABLE_INLINE_FUNCTION Real PowerMG::TemperatureFromDensityInternalEnergy(
+    const Real rho, const Real sie, Real *lambda) const {
+  Real value = (sie - AllHugInternalEnergyFromDensity(rho)) / _Cv0 +
+               AllHugTemperatureFromDensity(rho);
+  if (value < 0.0) {
+#ifndef NDEBUG
+    PORTABLE_WARN("Negative temperature");
+#endif // NDEBUG
+    value = 1.e-12;
+  }
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::PressureFromDensityInternalEnergy(
+    const Real rho, const Real sie, Real *lambda) const {
+  Real value = AllHugPressureFromDensity(rho) +
+               _rho0 * _G0 * (sie - AllHugInternalEnergyFromDensity(rho));
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::MinInternalEnergyFromDensity(const Real rho,
+                                                                    Real *lambda) const {
+  MinInternalEnergyIsNotEnabled("PowerMG");
+  return 0.0;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::EntropyFromDensityInternalEnergy(
+    const Real rho, const Real sie, Real *lambda) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real value = std::log(TemperatureFromDensityInternalEnergy(rho, sie) / _T0);
+  value = _S0 - _G0 * _Cv0 * eta + _Cv0 * value;
+  if (value < 0.0) {
+#ifndef NDEBUG
+    PORTABLE_WARN("Negative entropy");
+#endif // NDEBUG
+    value = 1.e-12;
+  }
+  return value;
+}
+// AEM: Give error since function is not well defined
+PORTABLE_INLINE_FUNCTION void
+PowerMG::DensityEnergyFromPressureTemperature(const Real press, const Real temp,
+                                              Real *lambda, Real &rho, Real &sie) const {
+  EOS_ERROR("PowerMG::DensityEnergyFromPressureTemperature: "
+            "Not implemented.\n");
+}
+// AEM: We should add entropy and Gruneissen parameters here so that it is complete
+// If we add also alpha and BT, those should also be in here.
```

Comment:
:+1: this is a good suggestion

Reply:
But as usual, something that needs to be changed in many places...

Reply:
Yeah---this is something for later as it's a relatively big refactor.